slackbot/dispatcher.py

```diff
@@ -14,33 +14,65 @@
+        self._pool = WorkerPool(self.dispatch)
         self._plugins = plugins
 
     def start(self):
         self._pool.start()
 
-    def dispatch_msg(self, msg):
-        category = msg[0]
-        msg = msg[1]
-        text = msg['text']
-        responded = False
-        for func, args in self._plugins.get_plugins(category, text):
+    def _execute_plugin(self, category, key, params):
+        executed = False
+        exception = None
+        for func, args in self._plugins.get_plugins(category, key):
             if func:
-                responded = True
+                executed = True
                 try:
-                    func(Message(self._client, msg), *args)
+                    func(params, *args)
                 except:
-                    logger.exception('failed to handle message %s with plugin "%s"', text, func.__name__)
-                    reply = '[%s] I have problem when handling "%s"\n' % (func.__name__, text)
-                    reply += '```\n%s\n```' % traceback.format_exc()
-                    self._client.rtm_send_message(msg['channel'], reply)
+                    logger.exception('Failed to handle message %s with plugin'
+                                     ' "%s"', key, func.__name__)
+                    exception = ('[%s] I have problem when handling "%s"\n' %
+                                 (func.__name__, key))
+                    exception += '```\n%s\n```' % traceback.format_exc()
+        return executed, exception
+
+    def dispatch(self, event):
+        """
+        Should dispatch all categories.
+        """
+        category, payload = event
+
+        logger.debug('dispatching category: %s, payload: %s' %
+                     (category, str(payload)))
+
+        if category in ['listen_to', 'respond_to']:
+            text = payload['text']
+            message = Message(self._client, payload)
+            responded, exception = self._execute_plugin(category, text, message)
+            if exception is not None:
+                self._client.rtm_send_message(payload['channel'], exception)
+            if not responded and category == 'responded_to':
+                self._default_reply(payload)
+        elif category == 'on_reaction':
+            reaction = payload['reaction']
+            user = payload['user']
+            reaction_event = payload['type'] # added/removed
+            item = payload['item']
+            timestamp = item['ts']
+            channel = item['channel']
+            item_type = item['type']
+
+            # Note that we can only find messages in public channels.
+            message = self._client.get_message(channel=channel, timestamp=timestamp)
+            params = { 'user' : user, 'type': reaction_event, 'message' : message }
+            executed, exception = self._execute_plugin(category, reaction, params)
+
+            assert executed, "Couldn't find any plugin for %s." % category
+            assert exception is None, exception
```

Comment:
No need to return a pair and assert them. It doesn't make sense to catch the exception in `_execute_plugin` and assert it here.